File: mailferry/imap/parser.py
"""
MailFerry - IMAP Migration & Sync
High-Performance Native IMAP Migration Engine

Copyright (C) 2026 Andy Saputra <andy@saputra.org>

https://saputra.org
https://github.com/ajsap/mailferry

Licensed under the GNU Affero General Public License v3.0 (AGPL-3.0).
This program is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

Contributions welcome: submit issues, feature requests and pull requests
at https://github.com/ajsap/mailferry

Incremental IMAP response tokenizer.

A logical server response is a sequence of *segments*: (text, literal) pairs,
where `text` is the line content with any trailing "{N}" marker stripped and
`literal` is the N raw bytes that followed it (None for the final segment).

Tokens: str atoms / quoted strings (latin-1 decoded, lossless), bytes for
literals, nested lists for parenthesised lists, None for NIL.
Atoms containing '[' consume through the matching ']' (BODY[HEADER.FIELDS
(...)] stays one token).
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional, Tuple

from ..errors import ProtocolError

Segment = Tuple[bytes, Optional[bytes]]

LITERAL_TAIL = re.compile(rb"\{(\d+)(\+)?\}$")
_STATUS_WORDS = {"OK", "NO", "BAD", "BYE", "PREAUTH"}


@dataclass
class Response:
    kind: str                     # 'tagged' | 'untagged' | 'cont'
    tag: str = ""
    status: str = ""              # OK/NO/BAD/BYE/PREAUTH for status responses
    dtype: str = ""               # untagged data type: FETCH/LIST/SEARCH/...
    num: Optional[int] = None     # leading number for EXISTS/FETCH/...
    code: list = field(default_factory=list)   # tokens inside [...]
    text: str = ""
    tokens: list = field(default_factory=list)


class _Cursor:
    """Token reader across segments."""

    def __init__(self, segs: List[Segment], start: int = 0):
        self.segs = segs
        self.si = 0
        self.i = start

    def _text(self) -> bytes:
        return self.segs[self.si][0]

    def at_end_of_seg(self) -> bool:
        return self.i >= len(self._text())

    def read_tokens(self) -> list:
        out: list = []
        stack: List[list] = [out]
        while True:
            # segment exhausted -> emit literal (if any), move on
            if self.at_end_of_seg():
                lit = self.segs[self.si][1]
                if lit is not None:
                    stack[-1].append(lit)
                if self.si + 1 < len(self.segs):
                    self.si += 1
                    self.i = 0
                    continue
                break
            t = self._text()
            c = t[self.i:self.i + 1]
            if c == b" ":
                self.i += 1
            elif c == b"(":
                new: list = []
                stack[-1].append(new)
                stack.append(new)
                self.i += 1
            elif c == b")":
                if len(stack) > 1:
                    stack.pop()
                self.i += 1
            elif c == b'"':
                self.i += 1
                buf = bytearray()
                while True:
                    if self.at_end_of_seg():
                        raise ProtocolError("unterminated quoted string")
                    ch = t[self.i:self.i + 1]
                    if ch == b"\\" and self.i + 1 < len(t):
                        buf += t[self.i + 1:self.i + 2]
                        self.i += 2
                    elif ch == b'"':
                        self.i += 1
                        break
                    else:
                        buf += ch
                        self.i += 1
                stack[-1].append(buf.decode("latin-1"))
            else:
                # atom (may embed [...] with spaces/parens inside)
                start = self.i
                depth = 0
                while not self.at_end_of_seg():
                    ch = t[self.i:self.i + 1]
                    if depth == 0 and ch in b" ()":
                        break
                    if ch == b"[":
                        depth += 1
                    elif ch == b"]":
                        depth = max(0, depth - 1)
                    self.i += 1
                atom = t[start:self.i].decode("latin-1")
                if atom.upper() == "NIL":
                    stack[-1].append(None)
                else:
                    stack[-1].append(atom)
        return out


def _split_code_text(rest: bytes) -> Tuple[list, str]:
    """Peel a leading [response-code] off a status line remainder."""
    if rest.startswith(b"["):
        depth = 0
        for i, b in enumerate(rest):
            ch = bytes((b,))
            if ch == b"[":
                depth += 1
            elif ch == b"]":
                depth -= 1
                if depth == 0:
                    inner = rest[1:i]
                    text = rest[i + 1:].strip()
                    toks = _Cursor([(inner, None)]).read_tokens()
                    return toks, text.decode("latin-1")
        # unbalanced -> treat as text
    return [], rest.decode("latin-1")


def parse_response(segs: List[Segment]) -> Response:
    if not segs:
        raise ProtocolError("empty response")
    first = segs[0][0]
    if first.startswith(b"+"):
        return Response(kind="cont", text=first[1:].strip().decode("latin-1"))

    sp1 = first.find(b" ")
    if sp1 < 0:
        raise ProtocolError(f"malformed response line: {first[:80]!r}")
    tag = first[:sp1].decode("latin-1")
    rest = first[sp1 + 1:]
    sp2 = rest.find(b" ")
    word = (rest if sp2 < 0 else rest[:sp2]).decode("latin-1").upper()
    tail = b"" if sp2 < 0 else rest[sp2 + 1:]

    if tag == "*":
        if word.isdigit():
            num = int(word)
            sp3 = tail.find(b" ")
            dtype = (tail if sp3 < 0 else tail[:sp3]).decode("latin-1").upper()
            after = b"" if sp3 < 0 else tail[sp3 + 1:]
            r = Response(kind="untagged", dtype=dtype, num=num)
            if dtype == "FETCH":
                segs2 = [(after, segs[0][1])] + segs[1:]
                r.tokens = _Cursor(segs2).read_tokens()
            return r
        if word in _STATUS_WORDS:
            code, text = _split_code_text(tail)
            return Response(kind="untagged", dtype=word, status=word, code=code, text=text)
        segs2 = [(tail, segs[0][1])] + segs[1:]
        r = Response(kind="untagged", dtype=word)
        r.tokens = _Cursor(segs2).read_tokens()
        return r

    # tagged completion
    if word not in _STATUS_WORDS:
        raise ProtocolError(f"unexpected tagged word {word!r}")
    code, text = _split_code_text(tail)
    return Response(kind="tagged", tag=tag, status=word, code=code, text=text)


# --------------------------------------------------------------------------
# FETCH helpers
# --------------------------------------------------------------------------

def fetch_pairs(tokens: list) -> dict:
    """Flatten a FETCH att-list into {NAME: value}; BODY[...] -> 'BODY'."""
    if len(tokens) == 1 and isinstance(tokens[0], list):
        tokens = tokens[0]
    out = {}
    i = 0
    while i + 1 <= len(tokens) - 1:
        name = tokens[i]
        if not isinstance(name, str):
            i += 1
            continue
        key = name.upper()
        if key.startswith("BODY[") or key.startswith("BODY.PEEK["):
            key = "BODY"
        out[key] = tokens[i + 1]
        i += 2
    return out


def as_bytes(v) -> bytes:
    if isinstance(v, bytes):
        return v
    if isinstance(v, str):
        return v.encode("latin-1")
    if v is None:
        return b""
    raise ProtocolError(f"expected string/literal, got {type(v).__name__}")


def as_int(v, default=None):
    try:
        return int(v)
    except (TypeError, ValueError):
        return default


def flags_of(v) -> List[str]:
    if isinstance(v, list):
        return [x for x in v if isinstance(x, str)]
    return []

File: internal/imapx/parser.go
// MailFerry — IMAP Migration & Sync
// A High-Performance Native IMAP Migration Engine
//
// Copyright (C) 2026 Andy Saputra
// Author: Andy Saputra <andy@saputra.org>
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This file is part of MailFerry (https://github.com/ajsap/mailferry).
// Licensed under the GNU Affero General Public License v3.0 or later;
// see the LICENSE file for details.

package imapx

// IMAP response tokenizer. A logical line (small literals already inlined
// as []byte tokens by the reader) becomes a token tree:
//   string       atom / quoted string / number
//   []byte       literal
//   []any        parenthesised list
// plus the response envelope (tag, status/type, number, code, text).

import (
	"fmt"
	"strconv"
	"strings"
)

type Response struct {
	Tag    string // "*", "+" or the command tag
	Status string // OK NO BAD PREAUTH BYE (when a status response)
	Type   string // EXISTS FETCH LIST SEARCH STATUS CAPABILITY FLAGS RECENT EXPUNGE ...
	Num    uint32 // leading number for "* 12 EXISTS" / "* 3 FETCH"
	Tokens []any  // remaining tokens after envelope
	Code   []any  // [CODE ...] response code content
	Text   string // trailing human text
	Raw    string
}

func (r *Response) IsStatus() bool { return r.Status != "" }

// tokenize parses the token portion of a line. segs carries the pieces
// split at inlined literal boundaries; lits the literal bytes between them.
func tokenize(segs []string, lits [][]byte) []any {
	type frame struct{ items []any }
	stack := []*frame{{}}
	push := func(v any) { f := stack[len(stack)-1]; f.items = append(f.items, v) }
	for si, seg := range segs {
		i := 0
		for i < len(seg) {
			c := seg[i]
			switch {
			case c == ' ':
				i++
			case c == '(':
				stack = append(stack, &frame{})
				i++
			case c == ')':
				if len(stack) > 1 {
					f := stack[len(stack)-1]
					stack = stack[:len(stack)-1]
					push(f.items)
				}
				i++
			case c == '"':
				j := i + 1
				var b strings.Builder
				for j < len(seg) {
					if seg[j] == '\\' && j+1 < len(seg) {
						b.WriteByte(seg[j+1])
						j += 2
						continue
					}
					if seg[j] == '"' {
						break
					}
					b.WriteByte(seg[j])
					j++
				}
				push(b.String())
				i = j + 1
			case c == '[':
				// bracketed body-section atom, e.g. BODY[HEADER.FIELDS (...)]
				depth := 0
				j := i
				for j < len(seg) {
					if seg[j] == '[' {
						depth++
					} else if seg[j] == ']' {
						depth--
						if depth == 0 {
							j++
							break
						}
					}
					j++
				}
				push(seg[i:j])
				i = j
			default:
				j := i
				for j < len(seg) && seg[j] != ' ' && seg[j] != '(' && seg[j] != ')' {
					if seg[j] == '[' { // attach bracket to preceding atom (BODY[...])
						depth := 0
						for j < len(seg) {
							if seg[j] == '[' {
								depth++
							} else if seg[j] == ']' {
								depth--
								if depth == 0 {
									j++
									break
								}
							}
							j++
						}
						continue
					}
					j++
				}
				push(seg[i:j])
				i = j
			}
		}
		if si < len(lits) {
			push(lits[si])
		}
	}
	return stack[0].items
}

var statusWords = map[string]bool{
	"OK": true, "NO": true, "BAD": true, "PREAUTH": true, "BYE": true,
}

// ParseLine builds a Response from a logical line.
func ParseLine(segs []string, lits [][]byte) (*Response, error) {
	if len(segs) == 0 {
		return nil, fmt.Errorf("empty response")
	}
	first := segs[0]
	r := &Response{Raw: first}
	sp := strings.IndexByte(first, ' ')
	if sp < 0 {
		if first == "+" {
			return &Response{Tag: "+"}, nil
		}
		return nil, fmt.Errorf("malformed response line: %q", first)
	}
	r.Tag = first[:sp]
	rest := first[sp+1:]
	if r.Tag == "+" {
		r.Text = rest
		return r, nil
	}

	// optional leading number: "* 12 EXISTS"
	sp2 := strings.IndexByte(rest, ' ')
	head := rest
	if sp2 >= 0 {
		head = rest[:sp2]
	}
	if n, err := strconv.ParseUint(head, 10, 32); err == nil && r.Tag == "*" {
		r.Num = uint32(n)
		if sp2 < 0 {
			return r, nil
		}
		rest = rest[sp2+1:]
		sp2 = strings.IndexByte(rest, ' ')
		head = rest
		if sp2 >= 0 {
			head = rest[:sp2]
		}
	}

	word := strings.ToUpper(head)
	if statusWords[word] {
		r.Status = word
		text := ""
		if sp2 >= 0 {
			text = rest[sp2+1:]
		}
		// [CODE args] prefix
		if strings.HasPrefix(text, "[") {
			if end := strings.IndexByte(text, ']'); end > 0 {
				code := text[1:end]
				r.Code = tokenize([]string{code}, nil)
				text = strings.TrimPrefix(text[end+1:], " ")
			}
		}
		r.Text = text
		return r, nil
	}

	r.Type = word
	tokRest := ""
	if sp2 >= 0 {
		tokRest = rest[sp2+1:]
	}
	all := append([]string{tokRest}, segs[1:]...)
	r.Tokens = tokenize(all, lits)
	return r, nil
}

// Helpers over token trees.

func TokStr(v any) (string, bool) {
	switch x := v.(type) {
	case string:
		return x, true
	case []byte:
		return string(x), true
	}
	return "", false
}

func TokUint(v any) (uint32, bool) {
	if s, ok := TokStr(v); ok {
		if n, err := strconv.ParseUint(s, 10, 32); err == nil {
			return uint32(n), true
		}
	}
	return 0, false
}

func TokInt64(v any) (int64, bool) {
	if s, ok := TokStr(v); ok {
		if n, err := strconv.ParseInt(s, 10, 64); err == nil {
			return n, true
		}
	}
	return 0, false
}

// FetchAttrs flattens a "* n FETCH (K V K V ...)" token list into a map
// keyed by upper-cased attribute name.
func FetchAttrs(tokens []any) map[string]any {
	out := map[string]any{}
	var list []any
	for _, t := range tokens {
		if l, ok := t.([]any); ok {
			list = l
			break
		}
	}
	if list == nil {
		list = tokens
	}
	i := 0
	for i+1 < len(list) {
		k, ok := TokStr(list[i])
		if !ok {
			i++
			continue
		}
		key := strings.ToUpper(k)
		if idx := strings.IndexByte(key, '['); idx > 0 {
			key = key[:idx] + "[]"
		}
		out[key] = list[i+1]
		i += 2
	}
	return out
}
